imports.py

```diff
@@ -0,0 +1,201 @@
+"""Imports for Jupyter Notebook"""
+import os
+import json
+from typing import Dict, Union
+
+import httpx
+from IPython.display import display, HTML
+from httpx import Response
+
+
+BASE_URL = "http://20.119.216.32:8000/r4/"  # TIMS server
+REQUEST_HEADERS = {
+	"Content-Type": "application/fhir+json; charset=UTF-8",
+	"Accept-Charset": "utf-8",
+	"User-Agent": "HAPI-FHIR/5.0.0 (FHIR Client; FHIR 4.0.1/R4; apache)"}
+PROJECT_DIR = os.path.dirname(__file__)
+CACHE_DIR = os.path.join(PROJECT_DIR, 'cache')
+URI = str
+
+
+# Utils
+def encode_filename_str(x):
+	"""Convert such that can create friendly system filename"""
+	delim = '*'
+	return x.replace('/', f'{delim}slash')
+
+
+def decode_filename_str(x):
+	"""Convert from friendly system filename"""
+	delim = '*'
+	return x.replace(f'{delim}slash', '/')
+
+
+def url_and_params_to_filename(url: str, params: Dict) -> str:
+	"""Convert to filename"""
+	x = url
+	if params:
+		x += '___params_'
+		for k, v in params.items():
+			x += '__' + k + '_' + v
+	filename = encode_filename_str(x) + '.json'
+	return filename
+
+
+def cache_write(url: str, params: Dict, response: Dict):
```

Comment:
I added some functions for caching: `cache_read`, `cache_write`, `url_and_params_to_filename`, `decode_filename_str`, and `encode_filename_str`. I might end up refactoring those last to into the read/write funcs.

Reply:
The server must have been pretty slow. I was running off of a local server for dev and it was pretty snappy. I can see that changing the caching flag is pretty easy. A more fully developed caching strategy would have a way to invalidate the cache. One way to do that here would be to record the file system timestamps of the code and check to see if they changed. If so, throw a flag to refresh.

Reply:
That is a good idea for an improvement. I'll create an issue here just to remember, although it's not a high priority.

The server was a little slow for certain requests, but the even bigger reason for the caching is that the server crashed several times (I made some issues and tagged Shahim). It does auto-restart, but it takes a few minutes.